### PR TITLE
Add basic char*_t support for libc (partial WG14 N2653)

### PR DESCRIPTION
**Pull request by Febbe**
This PR implements a part of WG14 N2653:
 - Define C23 char8_t
 - Define C11 char16_t
 - Define C11 char32_t
 
 Missing goals are:
    - The type of UTF-8 character literals is changed from unsigned char to char8_t. (Since UTF-8 character literals already have type unsigned char, this is not a semantic change).
    - New mbrtoc8() and c8rtomb() functions declared in <uchar.h> enable conversions between multibyte characters and UTF-8.
    - A new ATOMIC_CHAR8_T_LOCK_FREE macro.
    - A new atomic_char8_t typedef name.

### CODE REVIEW

**Comment by Febbe** [comment]
@tahonermann I am asking for a review.

File: libc/config/baremetal/api.td [hunk 1]
```diff
@@ -85,5 +85,10 @@ def TimeAPI : PublicAPI<"time.h"> {
 }
 
 def UCharAPI : PublicAPI<"uchar.h"> {
-  let Types = ["mbstate_t"];
+  let Types = [
+    "mbstate_t",
```
Comment (michaelrj-google) [inline libc/config/baremetal/api.td:89]: you should also add these types to the `api.td` for `linux` if you're adding this header to the list of linux headers.

File: libc/include/llvm-libc-types/char16_t.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+//===-- Definition of clock_t type ----------------------------------------===//
```
Comment (michaelrj-google) [inline libc/include/llvm-libc-types/char16_t.h:1]: nit: this header needs to be fixed. Same for the other files.

File: libc/include/llvm-libc-types/char16_t.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_TYPES_CHAR8_T_H
```
Comment (michaelrj-google) [inline libc/include/llvm-libc-types/char16_t.h:9]: this header guard needs to be updated, same for char32

File: libc/include/llvm-libc-types/char32_t.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_TYPES_CHAR8_T_H
+#define LLVM_LIBC_TYPES_CHAR8_T_H
+
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
+#include <stdint.h>
```
Comment (lntue) [inline libc/include/llvm-libc-types/char32_t.h:13]: Use `#include "../llvm-libc-macros/stdint-macros.h"` and add corresponding dependency.  Similarly for `char16_t.h`.

File: libc/config/baremetal/api.td (hunk 1 above)
Comment (Febbe) [reply libc/config/baremetal/api.td:89]: What about windows, I couldn't find any header list. Does this mean, all are included by default, or none of them?

**Comment by Febbe** [comment]
Thank you for the feedback. I applied all of your changes and added uchar (and wchar) also to the list of included headers on linux-aarch, linux-arm, linux-riscv

File: libc/include/llvm-libc-types/CMakeLists.txt [hunk 5]
```diff
@@ -90,6 +90,19 @@ add_header(tcflag_t HDR tcflag_t.h)
 add_header(struct_termios HDR struct_termios.h DEPENDS .cc_t .speed_t .tcflag_t)
 add_header(__getoptargv_t HDR __getoptargv_t.h)
 add_header(wchar_t HDR wchar_t.h)
+add_header(char8_t HDR char8_t.h)
+add_header(
+    char16_t 
+  HDR 
+    char16_t.h
+  DEPENDS
+    libc.include.llvm-libc-macros.stdint_macros)
```
Comment (michaelrj-google) [inline libc/include/llvm-libc-types/CMakeLists.txt:99]: nit: closing parenthesis on its own line

File: libc/include/llvm-libc-types/char16_t.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
+//===-- Definition of char16_t type
```
Comment (michaelrj-google) [inline libc/include/llvm-libc-types/char16_t.h:1]: nit: fix the formatting on this header (you'll need to delete some `-`

**Review by michaelrj-google — COMMENTED** [review]
Overall looks good with some minor style nits

**Comment by Febbe** [comment]
Ok, now everything should be fixed. I can squash the commits together if that's desired. But I don't have merge rights, someone has to merge this.

**Comment by michaelrj-google** [comment]
Github automatically squashes, I can merge this for you.

**Pull request merged**